### ISSUE TITLE
Monge DataSet

### ISSUE THREAD

**Issue by kalanityL** [rodolfolotte/deep-learning#1]
Hi @rodolfolotte 
following our conversation here https://github.com/MarvinTeichmann/KittiSeg/issues/137

I used the link you gave me (http://vision.mas.ecp.fr/Personnel/teboul/data.php)

And collected some images+GT from there. 

First remark - the tagging is not so good.. (see vizualisation monge_78 and the windows on the roof) 
![monge_78 jpg_visu](https://user-images.githubusercontent.com/28892491/35918067-222eaed8-0c19-11e8-9708-b6f54650b1a1.png)

Second, the color codes don't seem to match on hypes/FacadeSeg_Adam_pool5_3_10000.json
"balcony" : [255, 0, 255], (pink) -- while on the GT it is 128,0,255 (purple-ish)
did you change something (recolor the GT or change the json color code) before training with these data? 

I am going to filter the images to correct GT only, and try a first train. 
I'll let you know

**Comment by rodolfolotte** [comment]
Yes yes, I've checked these tagging mismatching before, but I simply ignored them and used as such. 
Regarding the color, as the colors of door, sky and balcony are close to secondary colors, i defined them by myself as: door (orange:  255, 128, 0), sky (cyan: 0, 255, 255) and balcony (magenta: 255, 0, 255). 

I assumed it has no problem with the GT images, since the inference uses the argmax output to label the final image, so, it takes an arbitrary color.

**Comment by kalanityL** [comment]
I dont understand what you mean that you do regarding the colors. 
If you have 128,0,255 on your GT, but you don't define it in the list of colors of the json ... how could it learn anything about it? It will just ignore this part of the tagging won't it?

**Comment by kalanityL** [comment]
the issue is that sky and balcony are not detected. 
Which corresponds exactly to the colors that are "wrong" in the json file - by wrong I mean that the color code you put in the json does not correspond to the color that is actually on the GT. 

I believe sky and balcony are not detected... just because the net has no GT to learn about it. 

change the hype file to 
 "colors": {
    "roof" : [0, 0, 255],
    "sky" : [128, 255, 255],
    "wall" : [255, 255, 0],
    "balcony" : [128, 0, 255],
    "window" : [255, 0, 0],
    "background" : [0, 0, 0],
    "door" : [255, 128, 0],
    "shop" : [0, 255, 0]

where  "sky" : [128, 255, 255], and  "balcony" : [128, 0, 255], are the colors code on the GT 

that should work I believe.. 


I am trying on my side this training too. 
(but got  Overall Precision : nan, Overall Recall: nan, Overall F1 score : nan ... so I might have something else wrong on my side ;) ) I will see at the end of the training in a few minutes

**Comment by rodolfolotte** [comment]
Well, it is a good guess. I will try with those parameters! Let you know soon! ;)
thank you!

**Comment by kalanityL** [comment]
![monge_50](https://user-images.githubusercontent.com/28892491/35921665-ff6d9b06-0c23-11e8-90a6-a3e27ef24c56.png)
confusing... 
it will take me a little more than a few minutes to check out what's wrong.. :D 
let me know if it went better on your side ;)

**Comment by rodolfolotte** [comment]
@kalanityL you simply got the right thing (all classes assigned). 

Now, you see the colors assigned to the wrong feature, right? In order to correct this, sort the colors in `/inputs/facade_seg_input.py` and `segment.py` files. Make sure the order of colors are all the same in this files (regarding to the order of colors in hypes). Then, train again and do the inference!

**Comment by kalanityL** [comment]
Ah right, I didn't update the /inputs/facade_seg_input.py order. 
that will assign the proper color to what it finds.. 

But not only the color switch is wrong. Balcony and door are missing from my outputs ..

**Comment by rodolfolotte** [comment]
Thats is because the number of iterations... your xentropy probably is still high! Let the model be trained with 20k, 50k and 100k. See what happen! ;)

**Comment by kalanityL** [comment]
I don't believe. I test on the training images, so normally even 1K iterations should be enough.Not to be perfect. But it should find at least a "little something" for the door and balcony - (i mean, see the windows. He finds them well already - there is no training-related reason he wouldn't find the doors the same way... something must be missing somewhere. I am checking.

**Comment by kalanityL** [comment]
![monge_54](https://user-images.githubusercontent.com/28892491/35923035-da7193bc-0c27-11e8-8746-f693bf84c076.png)
(after updating /inputs/facade_seg_input.py)

**Comment by kalanityL** [comment]
wow...
so now I just switch the order on the colors (on the json, the input/facade_seg and on segment)
and did a new training
and got this.. 
![monge_5](https://user-images.githubusercontent.com/28892491/35926018-d7417dee-0c2f-11e8-82a2-25585ac43e78.png)


freaking out a bit. 

So maybe you're right. 
Will run a training on 10K iter and see.

**Comment by kalanityL** [comment]
So something very not intuitive to me happens.. it looks like it is learning one class after the other.. 

1K
![monge_5](https://user-images.githubusercontent.com/28892491/35928770-a76672bc-0c36-11e8-9bb3-1deb1cd0f0e9.png)


2K
![monge_5](https://user-images.githubusercontent.com/28892491/35928777-ac6d9d80-0c36-11e8-85b8-7d9f647e26d8.png)

3K
![monge_5](https://user-images.githubusercontent.com/28892491/35928789-b38e9948-0c36-11e8-8f33-ba386eb18ca8.png)

4K
![monge_5](https://user-images.githubusercontent.com/28892491/35928800-b97b639a-0c36-11e8-9455-58feb09a8fce.png)


will see how it continues. 

And apparently the order of the classes matters too.  
I can't find any idea why. Do you have a clue?

**Comment by rodolfolotte** [comment]
hehhehe... Thats wonderful!!! And yes... that s the right behaviour... the first iterations arr made on the first classes... then, it iterates over and over until all labels are sufficient trained (low entropy)... I had to leave my job earlier today... I left my computer training... so, as soon as i check my results tomorrow... I post it here! 

Cheers! Well done!

**Comment by kalanityL** [comment]
So... looks like I solved your bug ;)

7K
![monge_5](https://user-images.githubusercontent.com/28892491/35961258-af957c9c-0cb5-11e8-9147-ddfcefe946e0.png)

17K
![monge_5](https://user-images.githubusercontent.com/28892491/35961271-bc9db1c0-0cb5-11e8-898e-5d6d393398d6.png)

18K
![monge_5](https://user-images.githubusercontent.com/28892491/35961278-c507d714-0cb5-11e8-9f60-b234d5313bfd.png)



These were test on the training images. 

Here are some results on random google images I picked. 
Considering my training set is of 43 images, I guess we can say it is not so bad ;) 

20K iterations
![google_kitty](https://user-images.githubusercontent.com/28892491/35961444-69b0e490-0cb6-11e8-8fdb-b1a61dfc1620.png)


So. It looks like I solved your bug ;)

Thank you (very very very much) for the code :)

**Comment by rodolfolotte** [comment]
Yes, considering the solution of the two remaining classes, we now can say we have a good version for multi-class segmentation. :) 

I'm having problems in my server... cannot run the code! :/ Today I arrived here and the server was down... I need to make sure i got the same results, then, I will do the final commit!

Thanks @MarvinTeichmann,  @shivam-kotwalia, and @bendidi :)

Well done!

**Comment by kalanityL** [comment]
Sure, I let you check - enjoy :) 

4 things I'd like to share with you about it: 

- it seems that the order of the classes is *ridiculously* changing the amount of iterations needed. I get perfect results with 6K iterations instead of 17K by just changing the class order. 
I need to run more trainings to make sure it is not due to the initialization but that it is really related to the order of classes. If yes, we'll need to understand a pattern on what class to put first in the json (those with less/more objects maybe, or it is size-area related. I will study this). If you have an idea let me know

- I made some changes (minor) changes to automatically load the classes from the json (and keep order of the json) in input/segment and in segment.py (and also when copying the json in the model_files/hypes.json at the begin of the training). This way changing the classes needs to be done only once in the initial json. I can send you those changes next week if you want to add it

- I tried to tv-continue one training, and it "lost" half of his knowledge on the way. Need to see if is related to my changes or not. 

- I still get  "Overall Precision : nan, Overall Recall: nan, Overall F1 score : nan" all along the training. It is annoying. But I guess I will open another issue about this one ;)

**Comment by rodolfolotte** [comment]
_it seems that the order of the classes is ridiculously changing the amount of iterations needed. I get perfect results with 6K iterations instead of 17K by just changing the class order.
I need to run more trainings to make sure it is not due to the initialization but that it is really related to the order of classes. If yes, we'll need to understand a pattern on what class to put first in the json (those with less/more objects maybe, or it is size-area related. I will study this). If you have an idea let me know_
**I have noticed this behavior before, but did not investigate. When I was looking for a possible error regarding this 2 missing classes, I changed the order of classes, but I remember to get pretty similar results. I used to put background in last position, such as sky, roof, window, wall, balcony, door, shop, background. If the order of classes is in fact an issue, trying to figure if it is region or size related will definitely help**

_I made some changes (minor) changes to automatically load the classes from the json (and keep order of the json) in input/segment and in segment.py (and also when copying the json in the model_files/hypes.json at the begin of the training). This way changing the classes needs to be done only once in the initial json. I can send you those changes next week if you want to add it_
**Please, pull-request your changes, even if it's minor changes!**

_I tried to tv-continue one training, and it "lost" half of his knowledge on the way. Need to see if is related to my changes or not._
**I never used this routines by tensorvision. The continue.py is in my fork, but still did touch it. :/**

_I still get "Overall Precision : nan, Overall Recall: nan, Overall F1 score : nan" all along the training. It is annoying. But I guess I will open another issue about this one ;)_
**The evaluation for multiclass segmentation is one of the issue that still need to be checked.**

**Comment by kalanityL** [comment]
Cool :)
I want to make some more tests before I send you the changes - probably after week end. 

for the evaluation I've seen that @shivam-kotwalia has a push called "evaluation per epoch fixed" 
https://github.com/shivam-kotwalia/KittiSeg/commit/055f4dcbe285ef21113923c0f75dca25545e16cc
I will have a look and see if it could be solution for here. 

I will also make some tests and experiments with color changes. It will probably take few days - I'm having a lot of fun ;) facade is a very cool dataset for playing around

If I have any valuable conclusions I'll let you know.

**Comment by rodolfolotte** [comment]
Cool @kalanityL, Thank you! 
From my side, I will also do some experiments to analysis the model by the influence of Adam and SGD optmizer, few or many iterations, different datasets, so on!

Cheers, Rodolfo!

**Comment by rodolfolotte** [comment]
@kalanityL , I have fixed the `continue.py`. Seems i had removed it from my fork. Now, I added it again and fixed a bug regarding libraries import.
